apps/managedidentity/managedidentity.go

```diff
@@ -76,59 +79,154 @@ func (s systemAssignedValue) value() string { return string(s) }
+	// cacheAccessorMu *sync.RWMutex
+	httpClient ops.HTTPClient
+	MiType     ID
+	// pmanager manager 	// todo :  expose the manager from base.
+	// cacheAccessor   		cache.ExportReplace
+}
+
+type ClientOptions struct {
+	httpClient ops.HTTPClient
+}
+
+type AcquireTokenOptions struct {
+	Claims string
+}
+
+type ClientOption func(o *ClientOptions)
 
-func (w withClaimsOption) AcquireTokenOption()           {}
-func (w withHTTPClientOption) AcquireTokenOption()       {}
-func (w withHTTPClientOption) apply(opts *clientOptions) { opts.httpClient = w.HttpClient }
+type AcquireTokenOption func(o *AcquireTokenOptions)
 
 // WithClaims sets additional claims to request for the token, such as those required by conditional access policies.
 // Use this option when Azure AD returned a claims challenge for a prior request. The argument must be decoded.
 func WithClaims(claims string) AcquireTokenOption {
-	return withClaimsOption{Claims: claims}
+	return func(o *AcquireTokenOptions) {
+		o.Claims = claims
+	}
 }
 
 // WithHTTPClient allows for a custom HTTP client to be set.
-func WithHTTPClient(httpClient ops.HTTPClient) Option {
-	return withHTTPClientOption{HttpClient: httpClient}
+func WithHTTPClient(httpClient ops.HTTPClient) ClientOption {
+	return func(o *ClientOptions) {
+		o.httpClient = httpClient
+	}
 }
 
 // Client to be used to acquire tokens for managed identity.
 // ID: [SystemAssigned()], [ClientID("clientID")], [ResourceID("resourceID")], [ObjectID("objectID")]
 //
 // Options: [WithHTTPClient]
-func New(id ID, options ...Option) (Client, error) {
+func New(id ID, options ...ClientOption) (Client, error) {
 	fmt.Println("idType: ", id.value())
 
-	opts := clientOptions{
-		claims: "claims",
+	opts := ClientOptions{ // work on this side where
+		httpClient: shared.DefaultClient,
 	}
 
 	for _, option := range options {
-		option.apply(&opts)
+		option(&opts)
 	}
 
-	authInfo, err := authority.NewInfoFromAuthorityURI("authorityURI", true, false)
+	client := Client{ // TODO :: check for http client
+		MiType:     id,
+		httpClient: opts.httpClient,
+	}
+
+	return client, nil
+}
+
+func getTokenForURL(url *url.URL, httpClient ops.HTTPClient) (accesstokens.TokenResponse, error) {
+	req, err := http.NewRequest(http.MethodGet, url.String(), nil)
 	if err != nil {
-		return Client{}, err
+		return accesstokens.TokenResponse{}, err
 	}
+	req.Header.Add("Metadata", "true")
 
-	authParams := authority.NewAuthParams(id.value(), authInfo)
-	client := Client{ // Note: Hey, don't even THINK about making Base into *Base. See "design notes" in public.go and confidential.go
-		AuthParams:      authParams,
-		cacheAccessorMu: &sync.RWMutex{},
-		// manager:         storage.New(token),
-		// pmanager:        storage.NewPartitionedManager(token),
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return accesstokens.TokenResponse{}, err
 	}
 
-	return client, err
+	// Pull out response body
+	responseBytes, err := io.ReadAll(resp.Body)
```

Comment:
What if the response status isn't `OK`?

Reply:
Line 135 is a response to my comment but perhaps that comment was too vague. As written, when the response isn't OK, we return an error (string) indicating the status code. That's it--just the status code. We should at least return a [CallErr](https://pkg.go.dev/github.com/AzureAD/microsoft-authentication-library-for-go/apps/errors#CallErr) so the caller can inspect the response. It would also be good to include any body content in the error message so callers don't always have to inspect the response to figure out what went wrong.

Reply:
Thank you for this, i will update it with CallErr

Reply:
Thanks @chlowell, that makes sense. +1 to including the error message. 

For errors such as 404, 503 etc. we can only include the HttpStatus code (e.g. 503 - Service Not Available) and maybe a hint to the developer on what to do, e.g. "Managed Identity request failed with HTTP error code 404. Source: IMDS. Possible cause: Not running on an Azure host or the IMDS service is temporarily unavailable"

For 400 error (Bad Request), we need to make sure we extract the exact error message from the service. For example "Incorrect client ID". 

It is necessary to have tests for these 2 type of service errors.